pkg/agent/controller/l7visibility/l7visibility_controller.go

```diff
@@ -0,0 +1,509 @@
+	if err != nil {
+		klog.ErrorS(err, "could not list Pods for given namespace", "namespace", ns)
+		return nil
+	}
+	selectedPods = append(selectedPods, pods...)
+
+	// Only select the annotated Pods, as non annotated Pods are handled separately
+	for _, pod := range selectedPods {
+		_, ok := pod.Annotations[types.L7VisibilityAnnotationKey]
+		if ok {
+			AnnotatedPods = append(AnnotatedPods, pod)
+		}
+	}
+	return AnnotatedPods
+}
+
+func (l7c *L7VisibilityController) syncPod(podNN string) error {
+	podNamespace, podName := k8s.SplitNamespacedName(podNN)
+	pod, err := l7c.podLister.Pods(podNamespace).Get(podName)
+	if err != nil {
+		// Remove the TC flows if the Pod has been deleted
+		return l7c.removeTCFlow(podNN)
+	}
+
+	// TCmarkflow will only be removed if both the Pod and Namespace annotations
+	// do not contain the L7 annotation
+	annotationValue, ok := pod.Annotations[types.L7VisibilityAnnotationKey]
+	if !ok {
+		podNamespace, err := l7c.namespaceLister.Get(pod.Namespace)
+		if err != nil {
+			return l7c.removeTCFlow(podNN)
+		}
+		annotationValue, ok = podNamespace.Annotations[types.L7VisibilityAnnotationKey]
+		if !ok {
+			return l7c.removeTCFlow(podNN)
+		}
+	} else {
+		// Namespace is deleted with Pod containing L7 visibility annotation
+		_, err := l7c.namespaceLister.Get(pod.Namespace)
+		if err != nil {
+			return l7c.removeTCFlow(podNN)
+		}
+	}
+
+	// Check if the annotation value is one of the specified values
+	var direction v1alpha2.Direction
+	annotationValue = strings.ToLower(annotationValue)
+	switch annotationValue {
+	case "ingress":
+		direction = v1alpha2.DirectionIngress
+	case "egress":
+		direction = v1alpha2.DirectionEgress
+	case "both":
+		direction = v1alpha2.DirectionBoth
+	default:
+		return errInvalidAnnotation
+	}
+	podInterfaces := l7c.interfaceStore.GetContainerInterfacesByPod(pod.Name, pod.Namespace)
+	if len(podInterfaces) == 0 {
+		return errPodInterfaceNotFound
+	}
+	sourceOfPort := []uint32{uint32(podInterfaces[0].OFPort)}
+
+	// Start Suricata before starting traffic control mark flows
+	l7c.l7Reconciler.StartSuricataOnce()
+
+	tcName := fmt.Sprintf("tcl7-%s", podNN)
+	if err := l7c.ofClient.InstallTrafficControlMarkFlows(tcName, sourceOfPort, l7c.targetPort, direction, v1alpha2.ActionMirror); err != nil {
+		return err
+	}
+	l7c.addPodNNDirToMap(podNN, direction)
+	return nil
+}
+
+func (l7c *L7VisibilityController) addPodNNDirToMap(podNN string, direction v1alpha2.Direction) {
+	l7c.l7PodNNDirMapMut.Lock()
+	defer l7c.l7PodNNDirMapMut.Unlock()
+	l7c.L7PodNNDirectionMap[podNN] = direction
+}
+
+func (l7c *L7VisibilityController) deletePodNNFromMap(podNN string) {
+	l7c.l7PodNNDirMapMut.Lock()
+	defer l7c.l7PodNNDirMapMut.Unlock()
+	delete(l7c.L7PodNNDirectionMap, podNN)
+}
+
+func (l7c *L7VisibilityController) getTCNamefromPodNNDirMap(podNN string) string {
+	l7c.l7PodNNDirMapMut.RLock()
+	defer l7c.l7PodNNDirMapMut.RUnlock()
+	if _, ok := l7c.L7PodNNDirectionMap[podNN]; ok {
```

Comment:
It looks redundant to have L7 and NN in the name of the map, perhaps just name it podToDirection